src/sagemaker_server.cc

```diff
@@ -628,61 +628,168 @@ SagemakerAPIServer::SageMakerMMEHandleInfer(
   }
 }
 
+TRITONSERVER_Error*
+SagemakerAPIServer::SageMakerMMECheckUnloadedModelIsUnavailable(
+    const char* model_name, bool* is_model_unavailable)
+{
+  LOG_VERBOSE(1) << "Inside function to check if model is unavailable";
+
+  /* Use the RepositoryIndex API to check if the model state has become
+  UNAVAILABLE i.e. model is no longer in the 'in-the-process-of' being
+  UNLOADED. Consequently, the reason field should be 'unloaded'.*/
+  TRITONSERVER_Message* server_model_index_message_ = nullptr;
+  uint32_t ready_flag = 0;  // value of 1 should be set if only the 'ready'
+                            // models are required from the index. In this case,
+                            // we need all models.
+  TRITONSERVER_ServerModelIndex(
+      server_.get(), ready_flag, &server_model_index_message_);
+
+  std::shared_ptr<TRITONSERVER_Message> managed_msg(
+      server_model_index_message_,
+      [](TRITONSERVER_Message* msg) { TRITONSERVER_MessageDelete(msg); });
+
+  const char* index_buffer;
+  size_t index_byte_size;
+
+  /*Handle the return from this function correctly*/
+  TRITONSERVER_MessageSerializeToJson(
+      server_model_index_message_, &index_buffer, &index_byte_size);
+
+  /* Read into json buffer*/
+  triton::common::TritonJson::Value server_model_index_json;
+  server_model_index_json.Parse(index_buffer, index_byte_size);
+
+  const char* name;
+  const char* state;
+  const char* reason;
+  size_t name_len;
+  size_t state_len;
+  size_t reason_len;
+
+  for (size_t id = 0; id < server_model_index_json.ArraySize(); ++id) {
+    triton::common::TritonJson::Value index_json;
+    server_model_index_json.IndexAsObject(id, &index_json);
+
+    RETURN_IF_ERR(index_json.MemberAsString("name", &name, &name_len));
+
+    if (std::string(name) == std::string(model_name)) {
+      RETURN_IF_ERR(index_json.MemberAsString("state", &state, &state_len));
+
+      if (std::string(state) == UNLOAD_EXPECTED_STATE_) {
```

Comment:
I think you should do the opposite checking here, i.e. set `is_model_unavailable = true;` at the beginning and mark it false if you have seen the model NOT in expected state and reason. Because the index will report model with multiple versions in different entries, say there is model "A" with version 1, 2, depending on the timing, you may receive the following model index:
```
{
  name: "A",
  version: "1",
  state: UNLOADING,
  reason: ""
},
{
  name: "A",
  version: "2",
  state: UNAVAILABLE,
  reason: "unloaded"
}
```

Reply:
Updated the code to not re-use the unload_err pointer. 

Regarding the multiple version scenario - SM currently expects only one model version in a single model repository, and this logic currently holds fine on testing. I might consider updating this in a future PR, thank you for the suggestion. Even so, I've added a log statement that also includes the version number.